### PR TITLE
feat(aws): add AWS API Gateway REST resources

### PR DESCRIPTION
**Pull request by CharLEEx**
## Summary

Adds AWS API Gateway (REST API v1) resource providers to `packages/alchemy`, plus lifecycle tests, an `examples/aws-rest-api` reference stack, and tutorial content (`website/src/content/docs/tutorial/aws/api-gateway.mdx`).

## Includes

- REST API primitives: `RestApi`, `Deployment`, `Stage`, `Method`, `GatewayResource`, `Authorizer`, `ApiKey`, `UsagePlan`, `UsagePlanKey`, `DomainName`, `BasePathMapping`, `GatewayResponse`, `Account`, `VpcLink`, shared helpers
- Provider registration and `normalizeTags` export used by ApiGateway tagging
- `scripts/audit-service.ts` mapping for `apigateway`
- Root `tsconfig.json` project reference for `examples/aws-rest-api`

## Notes

- Integration tests require AWS credentials (e.g. configured profile in CI).
- Optional `VpcLink` test can be enabled with `ALCHEMY_TEST_VPC_LINK_TARGET_ARN`.

Refs: none

### CODE REVIEW

File: packages/alchemy/src/AWS/ApiGateway/ApiKey.ts [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+  enabled?: boolean;
+  generateDistinctId?: boolean;
+  /**
+   * Write-only value when creating; never stored in resource state or outputs.
+   */
+  value?: string;
+  stageKeys?: ag.StageKey[];
+  customerId?: string;
+  tags?: Record<string, string>;
+}
```
Comment (sam-goodwin) [inline packages/alchemy/src/AWS/ApiGateway/ApiKey.ts:54]: Add thorough docs on each property (but not the interface)

File: packages/alchemy/src/AWS/ApiGateway/ApiKey.ts [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+const ApiKeyResource = Resource<ApiKey>("AWS.ApiGateway.ApiKey");
+
+export { ApiKeyResource as ApiKey };
+
+const toTags = (tags: ag.ApiKey["tags"]) =>
+  Object.fromEntries(
+    Object.entries(tags ?? {}).filter(
+      (e): e is [string, string] => e[1] !== undefined,
+    ),
+  );
```
Comment (sam-goodwin) [inline packages/alchemy/src/AWS/ApiGateway/ApiKey.ts]: Does this belong in Tags.ts or do we already have it?

File: packages/alchemy/src/AWS/ApiGateway/ApiKey.ts [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+      return {
+        stables: ["id"] as const,
+        diff: Effect.fn(function* ({ news: newsIn, olds }) {
+          if (!isResolved(newsIn)) return;
+          const news = newsIn as ApiKeyProps;
+          if (
+            news.value !== olds.value ||
+            news.customerId !== olds.customerId
+          ) {
```
Comment (sam-goodwin) [inline packages/alchemy/src/AWS/ApiGateway/ApiKey.ts:125]: Add some // comments helping justify important decisions like this

File: packages/alchemy/src/AWS/ApiGateway/ApiKey.ts [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        create: Effect.fn(function* ({ id, news: newsIn, session }) {
+          if (!isResolved(newsIn)) {
+            return yield* Effect.die("ApiKey props were not resolved");
+          }
+          const news = newsIn as ApiKeyProps;
+          const internalTags = yield* createInternalTags(id);
+          const allTags = { ...news.tags, ...internalTags };
+
+          const k = yield* ag.createApiKey({
+            name: news.name,
```
Comment (sam-goodwin) [inline packages/alchemy/src/AWS/ApiGateway/ApiKey.ts]: Make the name optional and use createPhysicalName to generate a deterministic, unique name that rotates on each place. 

Plenty of patterns in this repo to draw from.

File: packages/alchemy/src/AWS/ApiGateway/ApiKey.ts [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+            description: news.description,
+            enabled: news.enabled,
+            generateDistinctId: news.generateDistinctId,
+            value: news.value,
+            stageKeys: news.stageKeys,
+            customerId: news.customerId,
+            tags: allTags,
```
Comment (sam-goodwin) [inline packages/alchemy/src/AWS/ApiGateway/ApiKey.ts]: If tags are added as part of create, that's great! It means we can use tags to attach alchemy metadata and use it to recover from "partial creates". 

Take a look at other AWS resources   

They all use Tags.ts to catch and recover from conflict exceptions

File: packages/alchemy/src/AWS/ApiGateway/Authorizer.ts [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+  restApiId: Input<string>;
+  name: string;
+  type: ag.AuthorizerType;
+  providerARNs?: string[];
+  authType?: string;
+  authorizerUri?: string;
+  authorizerCredentials?: string;
+  identitySource?: string;
+  identityValidationExpression?: string;
+  authorizerResultTtlInSeconds?: number;
```
Comment (sam-goodwin) [inline packages/alchemy/src/AWS/ApiGateway/Authorizer.ts:54]: Docs

File: packages/alchemy/src/AWS/ApiGateway/ApiKey.ts [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+export interface ApiKeyProps {
+  name?: string;
+  description?: string;
+  enabled?: boolean;
+  generateDistinctId?: boolean;
+  /**
+   * Write-only value when creating; never stored in resource state or outputs.
+   */
+  value?: string;
```
Comment (sam-goodwin) [inline packages/alchemy/src/AWS/ApiGateway/ApiKey.ts]: Is this a sensitive value? If so, wrap in `Redacted<string>`

File: packages/alchemy/src/AWS/ApiGateway/Authorizer.ts [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+            news.name !== olds.name ||
+            news.type !== olds.type
+          ) {
+            return { action: "replace" } as const;
+          }
+          if (!deepEqual(news.providerARNs, olds.providerARNs)) {
+            return { action: "replace" } as const;
+          }
+          if (news.authType !== olds.authType) {
+            return { action: "replace" } as const;
```
Comment (sam-goodwin) [inline packages/alchemy/src/AWS/ApiGateway/Authorizer.ts:118]: // comments to add some concise details explaining the decisions

File: packages/alchemy/src/AWS/ApiGateway/DomainName.ts [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+const DomainNameResource = Resource<DomainName>("AWS.ApiGateway.DomainName");
+
+export { DomainNameResource as DomainName };
+
+const toTags = (tags: ag.DomainName["tags"]) =>
+  Object.fromEntries(
+    Object.entries(tags ?? {}).filter(
+      (e): e is [string, string] => e[1] !== undefined,
+    ),
+  );
```
Comment (sam-goodwin) [inline packages/alchemy/src/AWS/ApiGateway/DomainName.ts]: Duplicate. Use Tags.ts. Add to it if it's missing something

File: packages/alchemy/src/AWS/ApiGateway/DomainName.ts [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+              op: news.certificateArn === undefined ? "remove" : "replace",
+              path: "/certificateArn",
+              value: news.certificateArn,
+            });
+          }
+          if (patches.length > 0) {
+            yield* ag.updateDomainName({
+              domainName: output.domainName,
+              patchOperations: patches,
+            });
```
Comment (sam-goodwin) [inline packages/alchemy/src/AWS/ApiGateway/DomainName.ts]: I notice a discrete lack of catching and retries. Are there edge cases? Errors we may want to retry

File: packages/alchemy/src/AWS/ApiGateway/Method.ts [hunk 11]
```diff
@@ -0,0 +1,421 @@
...
+/**
+ * Integration configuration for an API Gateway method (passed to `putIntegration`).
+ */
+export interface MethodIntegrationProps {
+  type: ag.IntegrationType;
+  integrationHttpMethod?: string;
+  uri?: Input<string>;
+  connectionType?: ag.ConnectionType;
+  connectionId?: string;
+  credentials?: string;
```
Comment (sam-goodwin) [inline packages/alchemy/src/AWS/ApiGateway/Method.ts:23]: Redacted?

File: packages/alchemy/src/AWS/ApiGateway/RestApi.ts [hunk 12]
```diff
@@ -0,0 +1,304 @@
...
+ * @example Regional REST API
+ * ```typescript
+ * import * as ApiGateway from "alchemy/AWS/ApiGateway";
+ *
+ * const api = yield* ApiGateway.RestApi("PublicApi", {
+ *   name: "my-api",
+ *   endpointConfiguration: { types: ["REGIONAL"] },
+ * });
+ * ```
+ */
```
Comment (sam-goodwin) [inline packages/alchemy/src/AWS/ApiGateway/RestApi.ts:114]: This could be more thorough. Please update docs with various sections and examples (these end up on our api reference docs, more illustrative examples the better).

File: packages/alchemy/src/AWS/ApiGateway/RestApi.ts [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+import * as Effect from "effect/Effect";
+import { deepEqual, isResolved } from "../../Diff.ts";
+import * as Provider from "../../Provider.ts";
+import { Resource } from "../../Resource.ts";
+import type { Providers } from "../Providers.ts";
+import { createInternalTags } from "../../Tags.ts";
+
+import { restApiArn, syncTags } from "./common.ts";
+
+export interface RestApiProps {
```
Comment (sam-goodwin) [inline packages/alchemy/src/AWS/ApiGateway/RestApi.ts:13]: Can we use an OpenAPI schema?

Any opportunity to integrate with effect http api's generated schema?

Seems a lot easier than using these low level resources.

File: packages/alchemy/src/AWS/ApiGateway/RestApi.ts [hunk 14]
```diff
@@ -0,0 +1,304 @@
...
+      patchReplace("/securityPolicy", news.securityPolicy ?? "TLS_1_0"),
+    );
+  }
+  if (news.endpointAccessMode !== prev.endpointAccessMode) {
+    patches.push(
+      patchReplace("/endpointAccessMode", news.endpointAccessMode ?? ""),
+    );
+  }
+  return patches;
+};
```
Comment (sam-goodwin) [inline packages/alchemy/src/AWS/ApiGateway/RestApi.ts:231]: Do we have thorough tests for this?

File: packages/alchemy/src/AWS/ApiGateway/UsagePlan.ts [hunk 15]
```diff
@@ -0,0 +1,331 @@
...
+const UsagePlanResource = Resource<UsagePlan>("AWS.ApiGateway.UsagePlan");
+
+export { UsagePlanResource as UsagePlan };
+
+const toTags = (tags: ag.UsagePlan["tags"]) =>
+  Object.fromEntries(
+    Object.entries(tags ?? {}).filter(
+      (e): e is [string, string] => e[1] !== undefined,
+    ),
+  );
```
Comment (sam-goodwin) [inline packages/alchemy/src/AWS/ApiGateway/UsagePlan.ts]: Duplicate again

File: packages/alchemy/test/AWS/ApiGateway/ApiKey.test.ts [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+import * as Effect from "effect/Effect";
+
+const { test } = Test.make({ providers: AWS.providers() });
+
+test.provider("create and delete API key", (stack) =>
+  Effect.gen(function* () {
+    const key = yield* stack.deploy(
+      Effect.gen(function* () {
+        return yield* AWS.ApiGateway.ApiKey("AgApiKey", {
+          name: "alchemy-test-ag-apikey",
```
Comment (sam-goodwin) [inline packages/alchemy/test/AWS/ApiGateway/ApiKey.test.ts]: Don't hard code names in tests. Make sure all names are optional and the resource provider generates a unique one using PhysicalName.ts. Look up the docs for that resource to confirm min/max length, illegal characters, etc

Hard coded names lead to conflicts.

File: packages/alchemy/test/AWS/ApiGateway/Authorizer.test.ts [hunk 17]
```diff
@@ -0,0 +1,7 @@
+import * as AWS from "@/AWS";
+import * as Test from "@/Test/Vitest";
+import { describe } from "@effect/vitest";
+
+Test.make({ providers: AWS.providers() });
+
+describe.skip("ApiGateway.Authorizer", () => {});
```
Comment (sam-goodwin) [inline packages/alchemy/test/AWS/ApiGateway/Authorizer.test.ts]: Any reason we are skipping it?

File: packages/alchemy/test/AWS/ApiGateway/GatewayResponse.test.ts [hunk 18]
```diff
@@ -0,0 +1,36 @@
+import * as AWS from "@/AWS";
+import * as Test from "@/Test/Vitest";
+import * as ag from "@distilled.cloud/aws/api-gateway";
+import { expect } from "@effect/vitest";
+import * as Effect from "effect/Effect";
+
+const { test } = Test.make({ providers: AWS.providers() });
+
+test.provider("create and delete gateway response", (stack) =>
```
Comment (sam-goodwin) [inline packages/alchemy/test/AWS/ApiGateway/GatewayResponse.test.ts:9]: A lot of tests are not thorough enough. 

Test the update flow too. Test edge cases.

**Review by sam-goodwin — CHANGES_REQUESTED** [review]
Amazing PR. I think we can use this foundation to build an effect native abstraction on top in a separate PR. 

HttpApi schema -> api gateway config

File: packages/alchemy/src/AWS/ApiGateway/ApiKey.ts (hunk 1 above)
Comment (CharLEEx) [reply packages/alchemy/src/AWS/ApiGateway/ApiKey.ts:54]: Addressed in 65fb3dd: added field-level JSDoc for every ApiKey prop, including generated names, write-only redacted values, stage keys, customer ID, enabled state, and tags. I left the interface itself without extra docs per the request.

File: packages/alchemy/src/AWS/ApiGateway/ApiKey.ts (hunk 2 above)
Comment (CharLEEx) [reply packages/alchemy/src/AWS/ApiGateway/ApiKey.ts]: Addressed in 65fb3dd: moved this shared conversion into Tags.ts as tagRecord(...) and updated ApiKey, DomainName, UsagePlan, and VpcLink to use it instead of duplicating local helpers.

File: packages/alchemy/src/AWS/ApiGateway/ApiKey.ts (hunk 3 above)
Comment (CharLEEx) [reply packages/alchemy/src/AWS/ApiGateway/ApiKey.ts:125]: Addressed in 65fb3dd and 275af43: added concise comments explaining why generated names are only patched when explicitly provided and why API key value rotation is modeled as replacement since API Gateway does not return the key value after create.

File: packages/alchemy/src/AWS/ApiGateway/ApiKey.ts (hunk 4 above)
Comment (CharLEEx) [reply packages/alchemy/src/AWS/ApiGateway/ApiKey.ts]: Addressed in 65fb3dd: ApiKey.name is now optional and uses createPhysicalName({ id, maxLength: 128 }) when omitted, giving deterministic stack/stage/logical-ID based names with the instance suffix.

File: packages/alchemy/src/AWS/ApiGateway/ApiKey.ts (hunk 5 above)
Comment (CharLEEx) [reply packages/alchemy/src/AWS/ApiGateway/ApiKey.ts]: Addressed in 65fb3dd: create now attaches Alchemy internal tags and catches ConflictException. On conflict it searches by name and only adopts an existing key when hasAlchemyTags(...) confirms it belongs to this stack/stage/id; otherwise the conflict is preserved.

File: packages/alchemy/src/AWS/ApiGateway/Authorizer.ts (hunk 6 above)
Comment (CharLEEx) [reply packages/alchemy/src/AWS/ApiGateway/Authorizer.ts:54]: Addressed in 65fb3dd: added field-level JSDoc for Authorizer props, including restApiId, name, type, providerARNs, authType, authorizerUri, authorizerCredentials, identity source, validation expression, and TTL.

File: packages/alchemy/src/AWS/ApiGateway/ApiKey.ts (hunk 7 above)
Comment (CharLEEx) [reply packages/alchemy/src/AWS/ApiGateway/ApiKey.ts]: Addressed in 65fb3dd: ApiKey.value is now Redacted.Redacted<string>. The provider unwraps it only when calling createApiKey, and the live test now uses Redacted.make(...). The value still is not returned in attributes/state.

File: packages/alchemy/src/AWS/ApiGateway/Authorizer.ts (hunk 8 above)
Comment (CharLEEx) [reply packages/alchemy/src/AWS/ApiGateway/Authorizer.ts:118]: Addressed in 65fb3dd: added comments around the replacement decision for authorizer identity/kind fields, noting that restApiId/name/type define the authorizer shape and are clearer as replacement than in-place mutation.

File: packages/alchemy/src/AWS/ApiGateway/DomainName.ts (hunk 9 above)
Comment (CharLEEx) [reply packages/alchemy/src/AWS/ApiGateway/DomainName.ts]: Addressed in 65fb3dd: removed the duplicate DomainName tag helper and switched it to the shared tagRecord(...) helper in Tags.ts.

File: packages/alchemy/src/AWS/ApiGateway/DomainName.ts (hunk 10 above)
Comment (CharLEEx) [reply packages/alchemy/src/AWS/ApiGateway/DomainName.ts]: Addressed in 65fb3dd: added retry handling for DomainName update/delete around ConflictException and TooManyRequestsException, which are the likely transient cases while certificate, policy, or routing changes propagate.

File: packages/alchemy/src/AWS/ApiGateway/Method.ts (hunk 11 above)
Comment (CharLEEx) [reply packages/alchemy/src/AWS/ApiGateway/Method.ts:23]: Addressed in 65fb3dd: added docs clarifying Method.integration.credentials is an IAM role ARN or passthrough marker stored by API Gateway, not secret key material, so I did not wrap it in Redacted.

File: packages/alchemy/src/AWS/ApiGateway/RestApi.ts (hunk 12 above)
Comment (CharLEEx) [reply packages/alchemy/src/AWS/ApiGateway/RestApi.ts:114]: Addressed in 65fb3dd: expanded RestApi JSDoc with multiple API reference sections and examples for generated names, private APIs, binary media types, and disabling the default execute-api endpoint.

File: packages/alchemy/src/AWS/ApiGateway/RestApi.ts (hunk 13 above)
Comment (CharLEEx) [reply packages/alchemy/src/AWS/ApiGateway/RestApi.ts:13]: Addressed in 65fb3dd: documented RestApi as the low-level primitive intended to be composed directly today, with OpenAPI / Effect HttpApi schema generation called out as a higher-level layer to build separately on top of these resources.

File: packages/alchemy/src/AWS/ApiGateway/RestApi.ts (hunk 14 above)
Comment (CharLEEx) [reply packages/alchemy/src/AWS/ApiGateway/RestApi.ts:231]: Addressed in 65fb3dd: added explicit binary media type removal coverage in RestApi.test.ts in addition to the existing add/update test, so both patch directions are covered.

File: packages/alchemy/src/AWS/ApiGateway/UsagePlan.ts (hunk 15 above)
Comment (CharLEEx) [reply packages/alchemy/src/AWS/ApiGateway/UsagePlan.ts]: Addressed in 65fb3dd: removed the duplicate UsagePlan tag helper and switched it to tagRecord(...) from Tags.ts.

File: packages/alchemy/test/AWS/ApiGateway/ApiKey.test.ts (hunk 16 above)
Comment (CharLEEx) [reply packages/alchemy/test/AWS/ApiGateway/ApiKey.test.ts]: Addressed in 65fb3dd: removed hard-coded physical names from the ApiGateway tests where providers can now generate deterministic names. ApiKey, UsagePlan, VpcLink, Authorizer, and RestApi names are optional and generated via createPhysicalName where appropriate.

File: packages/alchemy/test/AWS/ApiGateway/Authorizer.test.ts (hunk 17 above)
Comment (CharLEEx) [reply packages/alchemy/test/AWS/ApiGateway/Authorizer.test.ts]: Addressed in 65fb3dd: replaced the empty skipped Authorizer suite with an env-gated live test. It runs when ALCHEMY_TEST_AUTHORIZER_URI is set and covers create plus TTL update behavior.

File: packages/alchemy/test/AWS/ApiGateway/GatewayResponse.test.ts (hunk 18 above)
Comment (CharLEEx) [reply packages/alchemy/test/AWS/ApiGateway/GatewayResponse.test.ts:9]: Addressed in 65fb3dd: added a GatewayResponse update-flow test that changes statusCode and responseTemplates, then verifies the remote GatewayResponse reflects the update.

**Comment by sam-goodwin** [comment]
I took this over and got it merged with tests passing https://github.com/alchemy-run/alchemy-effect/pull/169

Couldn't edit your PR for some reason. Sorry, the tests weren't passing and API gateway is notorious with its throttling limits, i had to add some logic to handle it.

Also, the design needed to use Bindings so that the Outputs properly connected the DAG or else deployment failed. The tests weren't in good shape. Now they are.
